src/auth/src/credentials/external_account_sources/executable_sourced.rs

```diff
@@ -0,0 +1,449 @@
+    }
+}
+
+impl std::error::Error for ExecutionError {}
+
+impl Display for ExecutionError {
+    /// Formats the error message to include retryability and source.
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self {
+                code: Some(code),
+                message: Some(message),
+            } => write!(
+                f,
+                "{MSG}, response contains unsuccessful response, code=<{}>, message=<{}>",
+                code, message
+            ),
+            _ => write!(
+                f,
+                "{MSG}, response must include `code` and `message` fields when unsuccessful"
+            ),
+        }
+    }
+}
+
+const MSG: &str = "failed to read subject token";
+// default timeout is defined by AIP-4117
+const DEFAULT_TIMEOUT_SECS: Duration = Duration::from_secs(30);
+const ALLOW_EXECUTABLE_ENV: &str = "GOOGLE_EXTERNAL_ACCOUNT_ALLOW_EXECUTABLES";
+
+#[async_trait::async_trait]
+impl SubjectTokenProvider for ExecutableSourcedCredentials {
+    async fn subject_token(&self) -> Result<String> {
+        if let Some(output_file) = self.executable.output_file.clone() {
+            let token = Self::from_output_file(output_file).await;
+            if token.is_ok() {
+                return token;
+            }
+        }
+        let timeout = match self.executable.timeout_millis {
+            Some(timeout) => Duration::from_millis(timeout.into()),
+            None => DEFAULT_TIMEOUT_SECS,
+        };
+        let token = Self::from_command(self.executable.command.clone(), timeout).await?;
+        if token.is_empty() {
+            let msg = format!("{MSG}, subject token is empty");
+            return Err(CredentialsError::from_msg(false, msg));
+        }
+
+        Ok(token)
+    }
+}
+
+impl ExecutableSourcedCredentials {
+    async fn from_output_file(output_file: String) -> Result<String> {
+        let content = std::fs::read_to_string(output_file)
+            .map_err(|e| CredentialsError::from_source(false, e))?;
+
+        Self::parse_token(content)
+    }
+
+    /// See details on security reason on [executable sourced credentials].
+    /// [executable sourced credentials]: https://google.aip.dev/auth/4117#determining-the-subject-token-in-executable-sourced-credentials
+    async fn from_command(command: String, timeout: Duration) -> Result<String> {
+        // For security reasons, we need our consumers to set this environment variable to allow executables to be run.
+        let allow_executable = std::env::var(ALLOW_EXECUTABLE_ENV)
+            .ok()
+            .unwrap_or("0".to_string());
+        if allow_executable != "1" {
+            return Err(CredentialsError::from_msg(
+                false,
+                "executables need to be explicitly allowed (set GOOGLE_EXTERNAL_ACCOUNT_ALLOW_EXECUTABLES to '1') to run",
+            ));
+        }
+
+        let (command, args) = Self::split_command(command);
```

Comment:
can we split the command once and store it?

Reply:
Same issue as on the timeout, we are reusing the ExternalAccountConfig struct as both the struct representing the external account config file and internal config, which makes hard to have computed values. Will merge the PR and improve things here.

Ref: https://github.com/googleapis/google-cloud-rust/pull/2406